README.md
# SVG for Everybody

[![NPM Version][npm-img]][npm] [![Build Status][ci-img]][ci]

[SVG for Everybody] adds [SVG External Content] support to [all browsers].

To use it now, include the script in your document.

```html
<script src="/path/to/svg4everybody.js"></script>
<script>svg4everybody(); // run it now or whenever you are ready</script>
```

To support Internet Explorer 6-8, include the legacy script instead.

```html
<script src="/path/to/svg4everybody.legacy.js"></script>
<script>svg4everybody(); // run it now or whenever you are ready</script>
```

_As of v2.0.0, you must manually call `svg4everybody()`. If you are using an AMD/CommonJS dependency loader then you may call it within the callback closure._

IE 6-8 require you to put the script in the `<head>` in order to shiv `<svg>` and `<use>` elements. For best results in IE, set [X-UA-Compatible] to `ie=edge`. This can be done with a response header from the server or the following HTML in the `<head>`.

```html
<meta http-equiv="x-ua-compatible" content="ie=edge">
```

## Usage

Create an SVG image.

**map.svg:**
```html
<svg xmlns="http://www.w3.org/2000/svg">
	<symbol id="codepen" viewBox="0 0 64 64">
		<title>CodePen</title>
		<path etc.../>
	</symbol>
	<symbol id="youtube" viewBox="0 0 64 64">
		<title>YouTube</title>
		<path etc.../>
	</symbol>
	<symbol id="twitter" viewBox="0 0 64 64">
		<title>Twitter</title>
		<path etc.../>
	</symbol>
</svg>
```

This spritemap displays fine in **Chrome**, **Safari 7.1+**, **Firefox**, **Edge 13+**, **Opera**. [SVG for Everybody] polyfills the experience in **Safari 6**, **IE 6+**, and **Edge 12**.

```html
<svg role="img" title="CodePen">
	<use href="map.svg#codepen"/>
</svg>
<svg role="img" title="YouTube">
	<use href="map.svg#youtube"/>
</svg>
<svg role="img" title="Twitter">
	<use href="map.svg#twitter"/>
</svg>
```

*IE 6-8 requires a trailing slash `/` when using a self-closing `<use/>` element.*

![3 SVG logos](http://i.imgur.com/87Npdzn.png)

Browsers not supporting SVG fallback to images.

```html
<svg role="img" title="CodePen">
	<img src="map.svg.codepen.png">
</svg>
<svg role="img" title="YouTube">
	<img src="map.svg.youtube.png">
</svg>
<svg role="img" title="Twitter">
	<img src="map.svg.twitter.png">
</svg>
```

By default, fallback images point to a PNG file in the same location as the SVG, only with the `#` hash replaced by a `.` dot and then appended with a `.png` extension. If you want to change this behavior, you can define your own fallback.

```js
svg4everybody({
	fallback: function (src, svg, use) {
		// src: current href String 
		// svg: current SVG Element 
		// use: current USE Element 

		return 'fallback.png'; // ok, always return fallback.png
	}
});
```

All `<use>` elements that are descendants of an `<svg>` are checked for external content. If you want to change this behavior, you can define your own validator.

```js
svg4everybody({
	validate: function (src, svg, use) {
		// src: current href String 
		// svg: current SVG Element 
		// use: current USE Element 

		return true; // ok, everything is valid
	}
});
```

You can override whether the script polyfills External Content at all (`polyfill`), or whether SVG should even be used over fallback images (`nosvg`).

```js
svg4everybody({
	nosvg: true, // shiv <svg> and <use> elements and use image fallbacks
	polyfill: true // polyfill <use> elements for External Content
});
```

*Use of the `nosvg` option will requires you to use the legacy version of SVG for Everybody.*

## Implementation status

Modern browsers support external content in SVGs, *except Edge*. No frets; we can shim it.

| OS  | Browser           | SVG | External Content | Shimmed |
|:---:|-------------------|:---:|:----------------:|:-------:|
| *   | Chrome 21+        | ✔   | ✔                | —       |
| *   | Chrome 14-20      | ✔   | ✖                | ✔       |
| *   | Firefox 4+        | ✔   | ✔                | —       |
| *   | Opera 23+         | ✔   | ✔                | —       |
| *   | Opera Mini 8+     | ✔   | ✔                | —       |
| And | And. Browser 40+  | ✔   | ✔                | —       |
| And | And. Browser 4.1+ | ✔   | ✖                | ✔       |
| iOS | iOS 8.1+          | ✔   | ✔                | —       |
| iOS | iOS 6-7           | ✔   | ✖                | ✔       |
| OSX | Saf 7.1+          | ✔   | ✔                | —       |
| OSX | Saf 6             | ✔   | ✖                | ✔       |
| Win | Edge 13+          | ✔   | ✔                | —       |
| Win | Edge 12           | ✔   | ✖                | ✔       |
| Win | IE 9 - 11         | ✔   | ✖                | ✔       |
| Win | IE 6 - 8          | ✖   | ✖                | ✔       |

As you can see, **all major browsers support external content**.

We had been waiting on Edge, previously, but [David Storey], Edge’s project manager assured us that native support for external content in SVGs was high on their checklist. We would [track progress] and [vote for attention] to this issue. Then, just as I predicted...

> I have complete faith in the Microsoft Edge team and absolutely expect support to arrive within the next few months.
>
> — Jon Neal (August, 2015)

All of our [dreams came true].

## Readability and accessibility

SVGs are compelling to use for many reasons, and one of them is their ease of accessibility.

Within your spritemap, have each sprite use a `<title>` element to identify itself.

```html
<symbol id="codepen">
	<title>CodePen</title>
	<path etc.../>
</symbol>
```

When this sprite is used, its title will be read aloud in [JAWS](http://www.freedomscientific.com/products/fs/JAWS-product-page.asp) and [NVDA](http://www.nvaccess.org/). Then, within your document, each sprite may use a `title` attribute to identify itself.

```html
<svg title="CodePen">
	<use href="map.svg#codepen"/>
</svg>
```

That title will be read aloud in [VoiceOver](http://www.apple.com/accessibility/osx/voiceover/) and [NVDA](http://www.nvaccess.org/). At present, the `title` attribute is the only way to properly read aloud an SVG in VoiceOver. I’ll let you know if this changes.

All together, **use the `title` attribute in your document and the `title` element in your SVG**.

ARIA roles may also be used to provide even more information to assistive technology.

When a sprite is merely decoration, use `role="presentation"`.

```html
<a href="//twitter.com/jon_neal"><svg role="presentation">
	<use href="map.svg#twitter"/>
</svg> Find me on Twitter</a>
```

Otherwise, use `role="img"` and remember to add a description.

```html
<a href="//twitter.com/jon_neal"><svg title="Find me on Twitter" role="img">
	<use href="map.svg#twitter"/>
</svg></a>
```

### Further reading

- [Tips for creating accessible SVG](https://www.sitepoint.com/tips-accessible-svg/)
- [Using ARIA to enhance SVG accessibility](http://blog.paciellogroup.com/2013/12/using-aria-enhance-svg-accessibility/)
- [SVG symbol a good choice for icons](http://css-tricks.com/svg-symbol-good-choice-icons/)
- [Implementing inline SVG Icons](https://kartikprabhu.com/article/inline-svg-icons)

## Optimized SVGs

SVG files, especially exported from vector tools, often contain tons of unnecessary data such as editor metadata, comments, hidden elements, and other stuff that can be safely removed without affecting SVG rendering result.

Use a tool like [SVGO] to optimize SVG spritemaps.

```bash
$ [sudo] npm install -g svgo
$ svgo spritemap.svg
```

[ci]:      https://travis-ci.org/jonathantneal/svg4everybody
[ci-img]:  https://img.shields.io/travis/jonathantneal/svg4everybody.svg
[npm]:     https://www.npmjs.com/package/svg4everybody
[npm-img]: https://img.shields.io/npm/v/svg4everybody.svg

[all browsers]: http://caniuse.com/svg
[David Storey]: https://twitter.com/dstorey/status/626514631884804096
[dreams came true]: https://dev.windows.com/en-us/microsoft-edge/platform/changelog/desktop/10586/?compareWith=10240
[SVG External Content]: http://css-tricks.com/svg-sprites-use-better-icon-fonts/##Browser+Support
[SVG for Everybody]: https://github.com/jonathantneal/svg4everybody
[SVGO]: https://github.com/svg/svgo
[track progress]: http://dev.modern.ie/platform/status/svgexternalcontent/?filter=f3e0000bf&search=svg
[vote for attention]: https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/6263916-svg-external-content
[X-UA-Compatible]: http://www.modern.ie/en-us/performance/how-to-use-x-ua-compatible


CHANGELOG.md
## 2.1.8 (2017-04-18)

- Added: Support for [custom attribute on `use` elements](https://github.com/jonathantneal/svg4everybody/pull/155)

## 2.1.7 (2017-03-03)

- Updated: Fixing internal use tag in polyfill

## 2.1.6 (2017-03-01)

- Updated: Ensure the interval begins to run

## 2.1.5 (2017-02-28)

- Updated: Fix for infinite loop in validation with polyfill
- Updated: Fix for IFrame svg use issue in Edge

## 2.1.4 (2016-06-15)

- Added: Support for [nested `use` elements](https://github.com/jonathantneal/svg4everybody/pull/117)
- Updated: Test document

## 2.0.3 (2016-02-02)

- Added: No polyfilling of Edge 12.10547+
- Updated: Source with documentation
- Updated: Caching for fetched SVG elements
- Updated: Test document

## 2.0.2 (2016-02-02)

- Added: Use non-minified non-legacy version as main file
- Added: caching for elements
- Updated: project configuration and stricter linting
- Added: Test document
- Updated documentation

## 2.0.1 (2015-11-24)

- Fix infinite loop issue with invalid <use> tags
- Update documentation

## 2.0.0 (2015-08-01)

- Added: UMD pattern and `svg4everybody` method to activate shim
- Added: Grunt build process
- Added: Option to remove legacy code from Grunt
- Added: Option to manually shim SVG or External Content from the browser
- Added: Option to customize PNG fallback from the browser
- Updated: SVG External Content feature-deficient browser detection
- Updated: Exclusive activation on `<use>` children of `<svg>`
- Updated: Preservation of existing `viewbox` attribute for SVGs
- Updated: Preservation of existing `width` and `height` attributes for PNG fallbacks
- Updated: Many performance improvements for caching, AJAX, and polling

## 1.0.0 (2015-07-28)

- Initial release


Gruntfile.js
module.exports = function (grunt) {
	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),
		eslint: {
			gruntfile: {
				options: {
					rules: {
						camelcase: [0],
						'global-require': [0]
					}
				},
				files: {
					src: ['Gruntfile.js']
				}
			},
			buildfile: {
				options: {
					globals: ['LEGACY_SUPPORT', 'svg4everybody'],
					rules: {
						'no-magic-numbers': [0],
						'no-unused-vars': [0]
					}
				},
				files: {
					src: ['lib/svg4everybody.js']
				}
			}
		},
		jscs: {
			gruntfile: {
				options: {
					requireCamelCaseOrUpperCaseIdentifiers: null
				},
				files: {
					src: ['Gruntfile.js']
				}
			},
			buildfile: {
				files: {
					src: ['lib/svg4everybody.js']
				}
			}
		},
		uglify: {
			build: {
				files: {
					'dist/svg4everybody.js': ['dist/svg4everybody.legacy.js']
				},
				options: {
					beautify: {
						beautify: true,
						bracketize: true
					},
					compress: {
						global_defs: {
							LEGACY_SUPPORT: false
						}
					},
					mangle: false,
					preserveComments: 'some'
				}
			},
			buildmin: {
				files: {
					'dist/svg4everybody.min.js': ['dist/svg4everybody.legacy.js']
				},
				options: {
					compress: {
						global_defs: {
							LEGACY_SUPPORT: false
						}
					},
					mangle: true,
					preserveComments: false
				}
			},
			legacy: {
				files: {
					'dist/svg4everybody.legacy.js': ['dist/svg4everybody.legacy.js']
				},
				options: {
					beautify: {
						beautify: true,
						bracketize: true
					},
					compress: {
						global_defs: {
							LEGACY_SUPPORT: true
						}
					},
					mangle: false,
					preserveComments: 'some'
				}
			},
			legacymin: {
				files: {
					'dist/svg4everybody.legacy.min.js': ['dist/svg4everybody.legacy.js']
				},
				options: {
					compress: {
						global_defs: {
							LEGACY_SUPPORT: true
						}
					},
					mangle: true,
					preserveComments: false
				}
			}
		},
		umd: {
			build: {
				options: {
					src: 'lib/svg4everybody.js',
					dest: 'dist/svg4everybody.legacy.js',
					globalAlias: 'svg4everybody',
					objectToExport: 'svg4everybody'
				}
			}
		},
		watch:  {
			files: ['lib/svg4everybody.js'],
			tasks: ['test', 'build']
		}
	});

	require('load-grunt-tasks')(grunt);

	// npm run test
	grunt.registerTask('test', ['eslint', 'jscs']);

	// npm run build, grunt build
	grunt.registerTask('build', ['test', 'umd', 'uglify']);

	// npm run watch, grunt build:watch
	grunt.registerTask('build:watch', ['build', 'watch']);

	// grunt
	grunt.registerTask('default', ['build']);
};


svg4everybody.ie8.js
/*! svg4everybody v1.1.0 | github.com/jonathantneal/svg4everybody */
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define(factory);
    } else if (typeof module === 'object' && module.exports) {
        // CommonJS
        module.exports = factory();
    } else {
        // Browser globals
        root.SVG4Everybody = factory();
    }
}(this, function (global) {

	var uses,
		requestAnimationFrame = requestAnimationFrame || setTimeout,
		CACHE = {},
		LTEIE8 = /MSIE\s[1-8]\b/.test(navigator.userAgent),
		IE9TO11 = /Trident\/[567]\b/.test(navigator.userAgent) || /Edge\/12/.test(navigator.userAgent) || (navigator.userAgent.match(/AppleWebKit\/(\d+)/) || [])[1] < 538;

	document.createElement('svg') && document.createElement('use');

	// methods
	function embed(svg, g) {
		if (g) {
			var
			viewBox = g.getAttribute('viewBox'),
			fragment = document.createDocumentFragment(),
			clone = g.cloneNode(true);

			if (viewBox) {
				svg.setAttribute('viewBox', viewBox);
			}

			while (clone.childNodes.length) {
				fragment.appendChild(clone.childNodes[0]);
			}

			svg.appendChild(fragment);
		}
	}

	function onload() {
		var xhr = this, x = document.createElement('x'), s = xhr.s;

		x.innerHTML = xhr.responseText;

		xhr.onload = function () {
			s.splice(0).map(function (array) {
				embed(array[0], x.querySelector('#' + array[1].replace(/(\W)/g, '\\$1')));
			});
		};

		xhr.onload();
	}

	function onframe() {
		var use;

		while ((use = uses[0])) {
			if (LTEIE8) {
				var img = new Image(), src, q;

				src = use.getAttribute('xlink:href');
				q = (/\?[^#]+/.exec(src) || [''])[0];
				img.src = src.replace(/\?[^#]+/, '').replace('#', '.').replace(/^\./, '') + '.png' + q;

				use.parentNode.replaceChild(img, use);
			} else {
				var
				svg = use.parentNode,
				url = use.getAttribute('xlink:href').split('#'),
				url_root = url[0],
				url_hash = url[1];

				svg.removeChild(use);

				if (url_root.length) {
					var xhr = CACHE[url_root] = CACHE[url_root] || new XMLHttpRequest();

					if (!xhr.s) {
						xhr.s = [];

						xhr.open('GET', url_root);

						xhr.onload = onload;

						xhr.send();
					}

					xhr.s.push([svg, url_hash]);

					if (xhr.readyState === 4) {
						xhr.onload();
					}

				} else {
					embed(svg, document.getElementById(url_hash));
				}
			}
		}

		requestAnimationFrame(onframe);
	}

	function SVG4Everybody() {
		if (LTEIE8 || IE9TO11) {
			uses = document.getElementsByTagName('use');
			onframe();
		}
	}

	// callable instance
	return SVG4Everybody;

}));


svg4everybody.w.events.js
/*! svg4everybody v1.0.0 | github.com/jonathantneal/svg4everybody */
/**
 * Modified version of the plugin with event handlers on
 * DOM attribute change
 */

/**
 * Define MutationObserver if not supported by the browser
 */
MutationObserver = (function () {
	if ( MutationObserver ) {
		return MutationObserver;
	}
	else {
		var
			MutationObserver = function ( cb_fn ) {
				this.cb_fn    = cb_fn;
				this.observed = [];

				setInterval( this.checkAttributes.bind(this), 1000 );
			};

		MutationObserver.prototype.disconnect = function ( el ) {
			var index;
			if ( index = this.isObserved( el ) ) {
				this.observed.splice( index, 1 );
			}
		};

		MutationObserver.prototype.observe = function ( el, configs ) {
			var config;
			if ( ! this.observed[ el ] && configs ) {
				for ( config in configs ) {
					if ( configs.hasOwnProperty( config ) ) {
						if ( config === 'attributes' ) {
							this.observed.push({
								el      : el,
								configs : configs,
								old     : this.getAttrValues( el )
							});
						}
						else {
							console.log( 'The property '+ config +' is not supported on this browser!' );
						}
					}
				}
			}
		};

		MutationObserver.prototype.isObserved = function ( el ) {
			for ( var i = 0, len = this.observed.length; i < len; i++ ) {
				if ( this.observed[ i ].el === el ) {
					break;
				}
			}
			return ( i != len ) ? i : false;
		};

		MutationObserver.prototype.getAttrValues = function ( el ) {
			var i, len, 
				attributes = el.attributes,
				attr_val   = {};

			for ( i = 0, len = attributes.length; i < len; i++ ) {
				attr = attributes[ i ];
				attr_val[ attr.nodeName ] = attr.nodeValue;
			}
			return attr_val;
		};

		MutationObserver.prototype.checkAttributes = function () {
			var i, len, el, attrValues, diff;
			for ( i = 0, len = this.observed.length; i < len; i++ ) {
				el = this.observed[ i ];
				attrValues = this.getAttrValues( el.el );
				diff = this.getDifferences( el.el, attrValues, el.old );

				if ( diff.length ) {
					this.cb_fn( diff );
					this.observed[i].old = attrValues;
				}
			}
		};

		MutationObserver.prototype.getDifferences = function ( el, now, old ) {
			var prop, 
				diff = [];

			for ( prop in now ) {
				if ( ! old.hasOwnProperty( prop ) || now[prop] !== old[prop] ) {
					diff.push({
						type          : 'attributes',
						target        : el,
						attributeName : prop,
						oldValue      : old[ prop ]
					});
				}
			}
			return diff;
		};

		return MutationObserver;
	}
})();

(function (document, uses, requestAnimationFrame, CACHE, IE9TO11) {
	var
		svgNS    = "http://www.w3.org/2000/svg",
		observer = new MutationObserver( function ( mutation ) {
			var i, len, target, parent, new_use, xlink;

			for ( i = 0, len = mutation.length; i < len; i++ ) {
				if ( mutation[ i ].type === 'attributes' &&
						 mutation[ i ].attributeName === 'xlink:href' ) {
					target = mutation[ i ].target;

					if ( xlink = target.getAttribute( 'xlink:href' ) ) {
						parent = target.parentNode;

						new_use = document.createElementNS( svgNS, 'use' );
						copyAttributes( target, new_use, [ 'id', 'xlink:href', 'x', 'y', 'transform' ] );
						parent.replaceChild( new_use, target  );
					}
				}
			}
		});

	function getAttributesList ( el ) {
		for (var i = 0, atts = el.attributes, n = atts.length, arr = []; i < n; i++){
	    arr.push(atts[i].nodeName);
		}
		return arr;
	}

	function copyAttributes ( from, to, attributes ) {
		var i, len, attr;

		attributes = attributes || getAttributesList( from );

		for ( i = 0, len = attributes.length; i < len; i++ ) {
			attr = attributes[ i ];
			to.setAttribute( attr, from.getAttribute( attr ) );
		}
	}

	function makeSVGgroup ( element ) {
		var g = document.createElementNS( svgNS, 'g' );
		g.appendChild( element );
		return g;
	}

	function getTransform ( element ) {
    var matrix = element.getAttribute( "transform" );
    if ( matrix == undefined ) {
      matrix = [ 1.0, 0.0, 0.0, 1.0, 0.0, 0.0];
    }
    else {
      if ( matrix.match ( 'matrix' ) ) {
        matrix = matrix.slice(7,-1).split(' ');
        for ( var i = 0; i < matrix.length; i++ ) {
          matrix[ i ] = parseFloat( matrix[ i ] );
        }
      }
      else if ( matrix.match ( 'translate' ) ) {
        matrix = matrix.slice( 10, -1 ).split(' ');
        matrix = [ 1.0, 0.0, 0.0, 1.0, matrix [ 0 ], matrix [ 1 ] ];
      }
    }
    
    return matrix;
  };
  
  function setTransform ( el, matrix ) {
    el.setAttribute("transform", "matrix(" + matrix.join(' ') + ")");
  };

	function fixPosition ( el, old ) {
		var
			x = old.getAttribute( 'x' ) * 1,
			y = old.getAttribute( 'y' ) * 1,
			matrix = getTransform( old );

			matrix[ 4 ] += x;
			matrix[ 5 ] += y;

			setTransform( el, matrix );
			el.setAttribute( 'x', 0 );
			el.setAttribute( 'y', 0 );
	}

	function embed(svg, g, old, place_holder, all ) {
		if (g) {
			var x, y,
			viewBox = g.getAttribute('viewBox'),
			fragment = document.createDocumentFragment(),
			clone = g.cloneNode(true);

			if (viewBox) {
				svg.setAttribute('viewBox', viewBox);
			}

			if ( all ) {
				fragment = clone;
			}
			else {
				while (clone.childNodes.length) {
					fragment.appendChild(clone.childNodes[0]);
				}
			}
			
			fragment = makeSVGgroup( fragment );
			copyAttributes( old, fragment );
			fixPosition( fragment, old );
			observer.observe( fragment, { attributes: true } );

			svg.replaceChild( fragment, place_holder );
		}
	}

	function onload() {
		var xhr = this, x = document.createElement('x'), s = xhr.s;

		x.innerHTML = xhr.responseText;

		xhr.onload = function () {
			s.splice(0).map(function (array) {
				embed(array[0], x.querySelector('#' + array[1].replace(/(\W)/g, '\\$1')), array[2], array[3]);
			});
		};

		xhr.onload();
	}

	function onframe() {
		var use;

		while ((use = uses[0])) {
			var
				g   = document.createElementNS( svgNS, 'g' ),
				svg = use.parentNode,
				url = use.getAttribute('xlink:href').split('#'),
				url_root = url[0],
				url_hash = url[1];

			observer.disconnect( use );
			svg.replaceChild( g, use );

			if (url_root.length) {
				var xhr = CACHE[url_root] = CACHE[url_root] || new XMLHttpRequest();

				if (!xhr.s) {
					xhr.s = [];

					xhr.open('GET', url_root);

					xhr.onload = onload;

					xhr.send();
				}

				xhr.s.push([svg, url_hash, use, g]);

				if (xhr.readyState === 4) {
					xhr.onload();
				}

			} else {
				embed(svg, document.getElementById(url_hash), use, g, true );
			}
		}

		requestAnimationFrame(onframe);
	}

	if (IE9TO11) {
		onframe();
	}
})(
	document,
	document.getElementsByTagName('use'),
	window.requestAnimationFrame || window.setTimeout,
	{},
	/Trident\/[567]\b/.test(navigator.userAgent) || (navigator.userAgent.match(/AppleWebKit\/(\d+)/) || [])[1] < 537
);



lib/svg4everybody.js
/*! svg4everybody v2.1.9 | github.com/jonathantneal/svg4everybody */

function embed(parent, svg, target, use) {
	// if the target exists
	if (target) {
		// create a document fragment to hold the contents of the target
		var fragment = document.createDocumentFragment();

		// cache the closest matching viewBox
		var viewBox = !svg.hasAttribute('viewBox') && target.getAttribute('viewBox');

		// conditionally set the viewBox on the svg
		if (viewBox) {
			svg.setAttribute('viewBox', viewBox);
		}

		// clone the target
		var clone = document.importNode ? document.importNode(target, true) : target.cloneNode(true);

		var g = document.createElementNS(svg.namespaceURI || 'http://www.w3.org/2000/svg', 'g');

		// copy the contents of the clone into the fragment
		while (clone.childNodes.length) {
			g.appendChild(clone.firstChild);
		}

		if (use) {
			for (var i = 0; use.attributes.length > i; i++) {
				var attr = use.attributes[i];
				if (attr.name === 'xlink:href' || attr.name === 'href') {
					continue;
				}
				g.setAttribute(attr.name, attr.value);
			}
		}

		fragment.appendChild(g);

		// append the fragment into the svg
		parent.appendChild(fragment);
	}
}

function loadreadystatechange(xhr, use) {
	// listen to changes in the request
	xhr.onreadystatechange = function () {
		// if the request is ready
		if (xhr.readyState === 4) {
			// get the cached html document
			var cachedDocument = xhr._cachedDocument;

			// ensure the cached html document based on the xhr response
			if (!cachedDocument) {
				cachedDocument = xhr._cachedDocument = document.implementation.createHTMLDocument('');

				cachedDocument.body.innerHTML = xhr.responseText;

				// ensure domains are the same, otherwise we'll have issues appending the
				// element in IE 11
				if (cachedDocument.domain !== document.domain) {
					cachedDocument.domain = document.domain;
				}

				xhr._cachedTarget = {};
			}

			// clear the xhr embeds list and embed each item
			xhr._embeds.splice(0).map(function (item) {
				// get the cached target
				var target = xhr._cachedTarget[item.id];

				// ensure the cached target
				if (!target) {
					target = xhr._cachedTarget[item.id] = cachedDocument.getElementById(item.id);
				}

				// embed the target into the svg
				embed(item.parent, item.svg, target, use);
			});
		}
	};

	// test the ready state change immediately
	xhr.onreadystatechange();
}

function svg4everybody(rawopts) {
	var opts = Object(rawopts);

	// create legacy support variables
	var nosvg;
	var fallback;

	// if running with legacy support
	if (LEGACY_SUPPORT) {
		// configure the fallback method
		fallback = opts.fallback || function (src) {
			return src.replace(/\?[^#]+/, '').replace('#', '.').replace(/^\./, '') + '.png' + (/\?[^#]+/.exec(src) || [''])[0];
		};

		// set whether to shiv <svg> and <use> elements and use image fallbacks
		nosvg = 'nosvg' in opts ? opts.nosvg : /\bMSIE [1-8]\b/.test(navigator.userAgent);

		// conditionally shiv <svg> and <use>
		if (nosvg) {
			document.createElement('svg');
			document.createElement('use');
		}
	}

	// set whether the polyfill will be activated or not
	var polyfill;
	var olderIEUA = /\bMSIE [1-8]\.0\b/;
	var newerIEUA = /\bTrident\/[567]\b|\bMSIE (?:9|10)\.0\b/;
	var webkitUA = /\bAppleWebKit\/(\d+)\b/;
	var olderEdgeUA = /\bEdge\/12\.(\d+)\b/;
	var edgeUA = /\bEdge\/.(\d+)\b/;
	//Checks whether iframed
	var inIframe = window.top !== window.self;

	if ('polyfill' in opts) {
		polyfill = opts.polyfill;
	} else if (LEGACY_SUPPORT) {
		polyfill = olderIEUA.test(navigator.userAgent) || newerIEUA.test(navigator.userAgent) || (navigator.userAgent.match(olderEdgeUA) || [])[1] < 10547 || (navigator.userAgent.match(webkitUA) || [])[1] < 537 || edgeUA.test(navigator.userAgent) && inIframe;
	} else {
		polyfill = newerIEUA.test(navigator.userAgent) || (navigator.userAgent.match(olderEdgeUA) || [])[1] < 10547 || (navigator.userAgent.match(webkitUA) || [])[1] < 537 || edgeUA.test(navigator.userAgent) && inIframe;
	}

	// create xhr requests object
	var requests = {};

	// use request animation frame or a timeout to search the dom for svgs
	var requestAnimationFrame = window.requestAnimationFrame || setTimeout;

	// get a live collection of use elements on the page
	var uses = document.getElementsByTagName('use');
	var numberOfSvgUseElementsToBypass = 0;

	function oninterval() {
		// if all <use>s in the array are being bypassed, don't proceed.
		if (numberOfSvgUseElementsToBypass && uses.length - numberOfSvgUseElementsToBypass <= 0) {
			return void requestAnimationFrame(oninterval, 67);
		}

		// if there are <use>s to process, proceed.

		// reset the bypass counter, since the counter will be incremented for every bypassed element,
		// even ones that were counted before.
		numberOfSvgUseElementsToBypass = 0;

		// get the cached <use> index
		var index = 0;

		// while the index exists in the live <use> collection
		while (index < uses.length) {
			// get the current <use>
			var use = uses[index];

			// get the current <svg>
			var parent = use.parentNode;
			var svg = getSVGAncestor(parent);
			var src = use.getAttribute('xlink:href') || use.getAttribute('href');

			if (!src && opts.attributeName) {
				src = use.getAttribute(opts.attributeName);
			}

			if (svg && src) {

				// if running with legacy support
				if (LEGACY_SUPPORT && nosvg) {
					// create a new fallback image
					var img = document.createElement('img');

					// force display in older IE
					img.style.cssText = 'display:inline-block;height:100%;width:100%';

					// set the fallback size using the svg size
					img.setAttribute('width', svg.getAttribute('width') || svg.clientWidth);
					img.setAttribute('height', svg.getAttribute('height') || svg.clientHeight);

					// set the fallback src
					img.src = fallback(src, svg, use);

					// replace the <use> with the fallback image
					parent.replaceChild(img, use);
				} else if (polyfill) {
					if (!opts.validate || opts.validate(src, svg, use)) {
						// remove the <use> element
						parent.removeChild(use);

						// parse the src and get the url and id
						var srcSplit = src.split('#');
						var url = srcSplit.shift();
						var id = srcSplit.join('#');

						// if the link is external
						if (url.length) {
							// get the cached xhr request
							var xhr = requests[url];

							// ensure the xhr request exists
							if (!xhr) {
								xhr = requests[url] = new XMLHttpRequest();

								xhr.open('GET', url);

								xhr.send();

								xhr._embeds = [];
							}

							// add the svg and id as an item to the xhr embeds list
							xhr._embeds.push({
								parent: parent,
								svg: svg,
								id: id
							});

							// prepare the xhr ready state change event
							loadreadystatechange(xhr, use);
						} else {
							// embed the local id into the svg
							embed(parent, svg, document.getElementById(id), use);
						}
					} else {
						// increase the index when the previous value was not "valid"
						++index;
						++numberOfSvgUseElementsToBypass;
					}
				}
			} else {
				// increase the index when the previous value was not "valid"
				++index;
			}
		}

		// continue the interval
		requestAnimationFrame(oninterval, 67);
	}

	// conditionally start the interval if the polyfill is active
	if (polyfill) {
		oninterval();
	}
}

function getSVGAncestor(node) {
	var svg = node;
	while (svg.nodeName.toLowerCase() !== 'svg') {
		svg = svg.parentNode;
		if (!svg) {
			break;
		}
	}
	return svg;
}
